### PR TITLE
[SYCL] Add ITT annotation instructions

### PR DESCRIPTION
**Pull request by MrSidims**
This patch introduces InstrumentalAnnotationsPass pass that
adds ITT instrumentations for barrier and barrier-like functions
and atomic instructions. The pass is being included to
compiler flow only when "-fsycl-instrument-device-code" option
passed to clang driver.

Following annotations were added:
- __itt_offload_wi_start_wrapper
  Notify tools work-item execution has started
- __itt_offload_wi_resume_wrapper
  Notify tools work-item execution resumed (e.g. after barrier)
- __itt_offload_wi_finish_wrapper
  Notify tools work-item execution has finished
- __itt_offload_wg_barrier
  Notify tools work-item has reached a barrier
- __itt_offload_atomic_op_start
   Atomic operation markup
- __itt_offload_atomic_op_finish
  Atomic operation markup

Signed-off-by: Dmitry Sidorov <dmitry.sidorov@intel.com>

### CODE REVIEW

**Comment by MrSidims** [comment]
- [ ] Resolve TODO https://github.com/intel/llvm/pull/3299/files#diff-79ddf74d80f0639bcdf2d93f0bbf8aa1bd46f18c2652d0fcfc092dac7d385018R146
- [ ] Add more tests?

`emitCall` function will be moved to a separate header in a separate patch

File: llvm/tools/sycl-post-link/InstrumentalAnnotations.cpp
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/InstrumentalAnnotations.cpp:197]: That is a lot of string comparisons, I wonder if it makes sense to try and optimize those.

A few ideas:
- we are only looking for functions containing `__spirv` - checking of this could be outlined into a separate early exit (`consume_front`)
- using `startswith` instead of `contains` should be more effective in negative cases and I assume that we are going to have tons of such cases (due to usage of other `__spirv_*` functions and presence of user-defined functions)
- I guess that at the beginning of the function name will be something like `_Z` + some numbers - in order to skip that you can take a look into `drop_while` or `ltrim` (probably would be more effective because of no function calls) `StringRef` methods

File: llvm/tools/sycl-post-link/InstrumentalAnnotations.cpp
Comment (kbobrovs) [inline llvm/tools/sycl-post-link/InstrumentalAnnotations.cpp:147]:
```suggestion
  // semantics of the instruction, encoded into a value of enum, defined like this
```

File: llvm/tools/sycl-post-link/InstrumentalAnnotations.cpp
Comment (kbobrovs) [inline llvm/tools/sycl-post-link/InstrumentalAnnotations.cpp:157]: Not sure this is a safe construct, as ArrayRef does not provide storage and just references the temp array on the right.

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (kbobrovs) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:113]: again, 'calls' seems too detailed for an option. `instrument`?

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (kbobrovs) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:809]: I'd expect the help message below to be extended with the new option

**Review by kbobrovs — COMMENTED** [review]
Please address this question first before getting to my other comments. From the first glance I don't see why this should be part of sycl-post-link.

> Is there a reason why this transformation needs to be done on linked code and can't be done on object code? Please state it here. Otherwise the transformation would have to go to BackendUtils.cpp or elsewhere.

File: llvm/tools/sycl-post-link/InstrumentalAnnotations.cpp
Comment (MichaelTutin) [reply llvm/tools/sycl-post-link/InstrumentalAnnotations.cpp:155]: It seems that combination of AcquireRelease semantic is missing in itt annotations. SequentiallyConsistent and release-consume semantics (if SPIR-V has some form of it) are not required for analysis at the moment.

I think we can either change this enum to a set of flags (acquire, release) or to add one more value to the enum. Another option would be to pass SPIR-V or C++ ordering value as is, but that will require extra argument to pass API name and tool will need to extract required details in runtime (a bit of extra overhead).
What is your opinion?

File: llvm/tools/sycl-post-link/InstrumentalAnnotations.cpp
Comment (MrSidims) [reply llvm/tools/sycl-post-link/InstrumentalAnnotations.cpp:9]: Tagging @romanovvlad  to check correctness of the answer.  From my understanding we first need to add definitions of these annotation instructions through libdevice (will be done https://github.com/intel/llvm/pull/3279 ).  I'll add a note about it here.

File: llvm/tools/sycl-post-link/InstrumentalAnnotations.cpp
Comment (MrSidims) [reply llvm/tools/sycl-post-link/InstrumentalAnnotations.cpp:155]: My vote goes to 
> to add one more value to the enum.

File: llvm/tools/sycl-post-link/InstrumentalAnnotations.cpp
Comment (romanovvlad) [reply llvm/tools/sycl-post-link/InstrumentalAnnotations.cpp:9]: @kbobrovs  Could you please clarify what "object code" means here?

Comment (kbobrovs) [reply llvm/tools/sycl-post-link/InstrumentalAnnotations.cpp:9]: > @kbobrovs Could you please clarify what "object code" means here?

I mean device LLVM IR within a fat object, which is not linked yet.

> From my understanding we first need to add definitions of these annotation instructions through libdevice (will be done #3279 ). I'll add a note about it here.

Definition is not needed to emit a call to function. The resulting code just needs to be linked later with a module containing definitions before converting to SPIRV. If that was the only reason, then this pass should be converted to Function pass and moved to BackendUtil.cpp.

**Comment by keryell** [comment]
Not clear what it is about. After some Googling, this seems to be related to Intel VTune: https://software.intel.com/content/www/us/en/develop/documentation/vtune-help/top/api-support/instrumentation-and-tracing-technology-apis/basic-usage-and-configuration/minimizing-itt-api-overhead.html
Is it something already used in Clang/LLVM upstream?
Otherwise some documentation is always appreciated. :-)
For example to add other markers, or to recycle the infrastructure for other instrumentation frameworks.

**Comment by bader** [comment]
> Not clear what it is about. After some Googling, this seems to be related to Intel VTune: https://software.intel.com/content/www/us/en/develop/documentation/vtune-help/top/api-support/instrumentation-and-tracing-technology-apis/basic-usage-and-configuration/minimizing-itt-api-overhead.html
> Is it something already used in Clang/LLVM upstream?
> Otherwise some documentation is always appreciated. :-)
> For example to add other markers, or to recycle the infrastructure for other instrumentation frameworks.

+1. Could folks working on this feature document the design in our internal documentation first, please? It's hard to follow changes in #3299 and #3279.

File: llvm/tools/sycl-post-link/InstrumentalAnnotations.cpp
Comment (MichaelTutin) [reply llvm/tools/sycl-post-link/InstrumentalAnnotations.cpp:155]: Ok. I've updated the document and ITT header file. The enum was changes to:
```C
enum __itt_atomic_mem_order_t
{
    __itt_mem_order_relaxed = 0,
    __itt_mem_order_acquire = 1,
    __itt_mem_order_release = 2,
    __itt_mem_order_acquire_release = 3
}
```

**Comment by MrSidims** [comment]
Push new patch set to unblock review. This comment is yet unaddressed: https://github.com/intel/llvm/pull/3299#discussion_r586768480 . Also there is an issue with passing of CodeGen option from the driver to BackendUtils.

**Comment by bader** [comment]
> Push new patch set to unblock review. This comment is yet unaddressed: [#3299 (comment)](https://github.com/intel/llvm/pull/3299#discussion_r586768480) . Also there is an issue with passing of CodeGen option from the driver to BackendUtils.

And there are merge conflicts with the `sycl` branch.

File: clang/lib/CodeGen/BackendUtil.cpp
Comment (MrSidims) [inline clang/lib/CodeGen/BackendUtil.cpp:842]: To revert this

File: clang/lib/CodeGen/BackendUtil.cpp
Comment (kbobrovs) [inline clang/lib/CodeGen/BackendUtil.cpp:981]: I guess there should be a check/error reporting somewhere when this option is used with incompatible target.

File: llvm/include/llvm/LinkAllPasses.h
Comment (kbobrovs) [inline llvm/include/llvm/LinkAllPasses.h:207]: please avoid redundant whitespace changes

File: llvm/tools/sycl-post-link/sycl-post-link.cpp
Comment (kbobrovs) [inline llvm/tools/sycl-post-link/sycl-post-link.cpp:788]: please remove unrelated change

File: clang/test/Driver/sycl-instrumentation.cpp
Comment (kbobrovs) [inline clang/test/Driver/sycl-instrumentation.cpp:6]: AFAIU, sycl post link is no longer touched. Should this test cases be updated? It still mentions -add-instrumentation-calls

**Review by kbobrovs — CHANGES_REQUESTED** [review]

File: llvm/include/llvm/LinkAllPasses.h
Comment (MrSidims) [reply llvm/include/llvm/LinkAllPasses.h:207]: I was a bit confused: should I follow general code style in the file, or should I follow surrounding SYCLlish style? So I decided to just align them, does it make sense?

Comment (kbobrovs) [reply llvm/include/llvm/LinkAllPasses.h:207]: In general, you should properly format only your changes, not touching existing. There can be exceptions, of course. @bader, do you think it is OK to in this case?

Comment (bader) [reply llvm/include/llvm/LinkAllPasses.h:207]: Ideally this change should go as a separate patch.

Comment (MrSidims) [reply llvm/include/llvm/LinkAllPasses.h:207]: okay

File: clang/test/Driver/sycl-instrumentation.cpp
Comment (MrSidims) [reply clang/test/Driver/sycl-instrumentation.cpp]: Indeed, fixed

File: clang/lib/CodeGen/BackendUtil.cpp
Comment (MrSidims) [reply clang/lib/CodeGen/BackendUtil.cpp]: Done, thanks!

**Review by romanovvlad — COMMENTED** [review]
LGTM

**Comment by vzakhari** [comment]
LGTM

File: llvm/include/llvm/Transforms/Instrumentation/SYCLITTAnnotations.h
Comment (bader) [inline llvm/include/llvm/Transforms/Instrumentation/SYCLITTAnnotations.h:10]:
```suggestion
// synchronization instructions. This can be used for kernel profiling.
```

File: llvm/lib/Transforms/Instrumentation/SYCLITTAnnotations.cpp
Comment (bader) [inline llvm/lib/Transforms/Instrumentation/SYCLITTAnnotations.cpp:22]:
```suggestion
/**
```
Let's avoid hardcoding specific library names. 
Feel free to add this to the design document instead.

File: llvm/include/llvm/Transforms/Instrumentation/SYCLITTAnnotations.h
Comment (bader) [inline llvm/include/llvm/Transforms/Instrumentation/SYCLITTAnnotations.h:1]: It would be great if you assign yourself as a code owner for this pass in the CODEOWNERS file.

File: llvm/lib/Transforms/Instrumentation/SYCLITTAnnotations.cpp
Comment (bader) [inline llvm/lib/Transforms/Instrumentation/SYCLITTAnnotations.cpp:143]:
```suggestion
// Check for calling convention of a function. If it's spir_kernel - consider
// the function to be a SYCL kernel.
bool isSPIRKernel(Function &F) {
```

Why is this pass SYCL specific? It should work for SPIR produced by other font-ends producing SPIR-V friendly LLVM IR. Right?
If you agree, please, replace SYCL with SPIR in file/pass names.

Comment (MrSidims) [reply llvm/lib/Transforms/Instrumentation/SYCLITTAnnotations.cpp:143]: You are right, thanks!

File: llvm/include/llvm/Transforms/Instrumentation/SPIRITTAnnotations.h
Comment (bader) [inline llvm/include/llvm/Transforms/Instrumentation/SPIRITTAnnotations.h:1]:
```suggestion
//===----- SPIRITTAnnotations.h - SPIR Instrumental Annotations Pass ------===//
```

File: llvm/include/llvm/Transforms/Instrumentation/SPIRITTAnnotations.h
Comment (bader) [inline llvm/include/llvm/Transforms/Instrumentation/SPIRITTAnnotations.h:9]:
```suggestion
// A transformation pass which adds instrumental calls to annotate SPIR
```

File: llvm/lib/Transforms/Instrumentation/SPIRITTAnnotations.cpp
Comment (bader) [inline llvm/lib/Transforms/Instrumentation/SPIRITTAnnotations.cpp:1]:
```suggestion
//===---- SPIRITTAnnotations.cpp - SPIR Instrumental Annotations Pass -----===//
```

File: llvm/lib/Transforms/Instrumentation/SPIRITTAnnotations.cpp
Comment (bader) [inline llvm/lib/Transforms/Instrumentation/SPIRITTAnnotations.cpp:9]:
```suggestion
// A transformation pass which adds instrumental calls to annotate SPIR
```

File: llvm/lib/Transforms/Instrumentation/SPIRITTAnnotations.cpp
Comment (bader) [inline llvm/lib/Transforms/Instrumentation/SPIRITTAnnotations.cpp:132]:
```suggestion
                "Insert ITT annotations for SPIR code", false, false)
```

File: llvm/lib/Transforms/Instrumentation/SPIRITTAnnotations.cpp
Comment (bader) [inline llvm/lib/Transforms/Instrumentation/SPIRITTAnnotations.cpp:142]:
```suggestion
// Check for calling convention of a function.
```

File: llvm/lib/Transforms/Instrumentation/SPIRITTAnnotations.cpp
Comment (bader) [inline llvm/lib/Transforms/Instrumentation/SPIRITTAnnotations.cpp:244]:
```suggestion
    // Annotate only SPIR kernels
```

**Review by bader — COMMENTED** [review]
Please, update comments accordingly.

**Review by bader — DISMISSED** [review]
I'm okay with LLVM pass.

**Review by mdtoguchi — DISMISSED** [review]

**Review by kbobrovs — DISMISSED** [review]

File: clang/test/CodeGenSYCL/kernel-simple-instrumentation.cpp
Comment (elizabethandrews) [inline clang/test/CodeGenSYCL/kernel-simple-instrumentation.cpp:9]: Please a short comment above the test to describe what it is testing.

File: clang/lib/CodeGen/BackendUtil.cpp
Comment (MrSidims) [reply clang/lib/CodeGen/BackendUtil.cpp]: Thanks!

**Review by elizabethandrews — APPROVED** [review]
LGTM. Thanks!

**Comment by MrSidims** [comment]
@bader can we consider the review to be done, as we have 3 LGTMs from the appropriate code owners and 3 from Vlad, Slava and Konst?

File: llvm/lib/Transforms/Instrumentation/SPIRITTAnnotations.cpp
Comment (AlexeySachkov) [inline llvm/lib/Transforms/Instrumentation/SPIRITTAnnotations.cpp:135]:
```suggestion
ModulePass *llvm::createSPIRITTAnnotationsLegacyPass() {
```

File: llvm/lib/Transforms/Instrumentation/SPIRITTAnnotations.cpp
Comment (AlexeySachkov) [inline llvm/lib/Transforms/Instrumentation/SPIRITTAnnotations.cpp:155]:
```suggestion
  return CallInst::Create(FT, FC.getCallee(), Args, "", InsertBefore);
```

File: llvm/lib/Transforms/Instrumentation/SPIRITTAnnotations.cpp
Comment (AlexeySachkov) [inline llvm/lib/Transforms/Instrumentation/SPIRITTAnnotations.cpp:166]: Do we really need to return that `true` here? Can we just make this function `void`?

File: llvm/lib/Transforms/Instrumentation/SPIRITTAnnotations.cpp
Comment (AlexeySachkov) [inline llvm/lib/Transforms/Instrumentation/SPIRITTAnnotations.cpp:197]: If you change your prefixes, you could use [`llvm::StringSwitch`](https://llvm.org/doxygen/classllvm_1_1StringSwitch.html) here: I mean use `_Z[number]__spirv_AtomicLoad` and `llvm::StringSwitch::StartsWith`

File: llvm/lib/Transforms/Instrumentation/SPIRITTAnnotations.cpp
Comment (AlexeySachkov) [inline llvm/lib/Transforms/Instrumentation/SPIRITTAnnotations.cpp:214]: Please note that in general this is not required to be a constant value.

Also, you are using `dyn_cast`, but don't check if it returned `nullptr`

**Review by AlexeySachkov — COMMENTED** [review]
Some post-commit comments: mostly minor ones, but the one about non-constant memory semantic might be important

**Pull request merged**
